luda-adventure/src/app/game/types/game_object/types/positioner/positioner.rs

```diff
@@ -0,0 +1,156 @@
+use super::{
+    get_heading_from_velocity, predict_collision, CollisionDirection, Heading, Movement,
+    MovementPlan, Velocity,
+};
+use crate::app::game::{Collider, TileExt};
+use namui::prelude::*;
+
+#[derive(ecs_macro::Component)]
+pub struct Positioner {
+    movement_plan: MovementPlan,
+    pub heading: Heading,
+}
+
+impl Positioner {
+    pub fn new(movement_plan: MovementPlan) -> Self {
+        Self {
+            movement_plan,
+            heading: Heading::Right,
+        }
+    }
+    pub fn set_velocity(&mut self, current_time: Time, velocity: Velocity, duration: Time) {
+        let current_position = self.get_position(current_time);
+        self.movement_plan =
+            MovementPlan::move_now(current_position, current_time, duration, velocity);
+        if let Some(heading) = get_heading_from_velocity(velocity) {
+            self.heading = heading;
+        }
+    }
+    pub fn get_predicted_movement_end_time(&self) -> Time {
+        self.movement_plan.get_predicted_movement_end_time()
+    }
+    pub fn predict_movement(
+        &mut self,
+        collider: &Collider,
+        collision_box_list: &[crate::app::game::CollisionBox],
+    ) {
+        let prediction_start_time = self.get_predicted_movement_end_time();
```

Comment:
```rust
let prediction_start_time = self.get_predicted_movement_end_time();
                                                      // ^-- `end` is right?
```

Reply:
Yes, next prediction should start from end of previous prediction.